pkg/pool-weighted/test/managed/BasePool.test.ts

```diff
@@ -0,0 +1,830 @@
+          toInternalBalance: false,
+        };
+
+        await expect(
+          vault.connect(sender).exitPool(poolId, sender.address, recipient.address, request)
+        ).to.be.revertedWith('NOT_IN_RECOVERY_MODE');
+      });
+
+      itSwaps();
+
+      itJoins();
+
+      itExits();
+    });
+
+    context('when in recovery mode', () => {
+      sharedBeforeEach('enable recovery mode', async () => {
+        const enableRecoveryAction = await actionId(pool, 'enableRecoveryMode');
+        const disableRecoveryAction = await actionId(pool, 'disableRecoveryMode');
+        await authorizer
+          .connect(admin)
+          .grantPermissions([enableRecoveryAction, disableRecoveryAction], admin.address, [ANY_ADDRESS, ANY_ADDRESS]);
+
+        await pool.connect(admin).enableRecoveryMode();
+      });
+
+      itSwaps();
+
+      itJoins();
+
+      itExits();
+
+      function itExitsViaRecoveryModeCorrectly() {
+        it('the recovery mode exit can be used', async () => {
+          const preExitBPT = await pool.balanceOf(sender.address);
+          const exitBPT = preExitBPT.div(3);
+
+          const request: ExitPoolRequest = {
+            assets: tokens.addresses,
+            minAmountsOut: Array(tokens.length).fill(0),
+            userData: defaultAbiCoder.encode(['uint256', 'uint256'], [RECOVERY_MODE_EXIT_KIND, exitBPT]),
+            toInternalBalance: false,
+          };
+
+          // The sole BPT holder is the owner, so they own the initial balances
+          const expectedChanges = tokens.reduce(
+            (changes, token, i) => ({ ...changes, [token.symbol]: ['very-near', initialBalances[i].div(3)] }),
+            {}
+          );
+          await expectBalanceChange(
+            () => vault.connect(sender).exitPool(poolId, sender.address, recipient.address, request),
+            tokens,
+            { account: recipient, changes: expectedChanges }
+          );
+
+          // Exit BPT was burned
+          const afterExitBalance = await pool.balanceOf(sender.address);
+          expect(afterExitBalance).to.equal(preExitBPT.sub(exitBPT));
+        });
+      }
+
+      itExitsViaRecoveryModeCorrectly();
+
+      context('when paused', () => {
+        sharedBeforeEach('pause pool', async () => {
+          await authorizer
+            .connect(admin)
+            .grantPermissions([await actionId(pool, 'pause')], admin.address, [ANY_ADDRESS]);
+
+          await pool.connect(admin).pause();
+        });
+
+        itExitsViaRecoveryModeCorrectly();
+      });
+    });
+
+    function itSwaps() {
+      describe('minimal swaps', () => {
+        let singleSwap: SingleSwap;
+
+        sharedBeforeEach(() => {
+          singleSwap = {
+            poolId: minimalPoolId,
+            kind: SwapKind.GivenIn,
+            assetIn: tokens.get(0).instance.address,
+            assetOut: tokens.get(1).instance.address,
+            amount: 1, // Needs to be > 0
+            userData: '0xdeadbeef',
```

Comment:
We don't need anything in particular in this field, but the test fails if this is set to `'0x'`.

This is because when we check the event's fields in `expectEvent.inIndirectReceipt` [here](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1805/files#diff-e5c00b397fb93defb4fa12d0ec28996b558552b4b2c72efb274347b17201c6d0R745), it ends up doing a `deep.eq` here: https://github.com/balancer-labs/balancer-v2-monorepo/blob/ad3ece6be2ee42e00cd64c52f4063676bdc65928/pvt/helpers/src/test/expectEvent.ts#L110

which in turn tries to convert every element of the object under comparison to a big number, and when it reaches `'0x'` it fails.

In short, this seems to be an issue with how TS interprets the contents and how `deep.eq` is implemented. I'm not 100% sure what a reasonable fix would be.

Reply:
Yes, our helper is not powerful enough to do this type of complex object comparison. Our options are a) make the helper better, but this will be hard and not very useful, or b) not pass `request` to the helper, which will make it not compare it, and instead use `inIndirectReceipt`'s return value (the event) to do a custom comparison yourself in your test.

Reply:
Thanks; I just wanted to raise this up in case I was missing anything.

For the purpose of this test, this value is just as good as any other (maybe even better than an empty `'0x'`). So I would leave it like this since it also simplifies the test.